### ISSUE TITLE
CUDA implementation of BiasAddGrad op is non-determinstic

### ISSUE THREAD

**Issue by eamartin** [tensorflow/tensorflow#22398]
I'm running TensorFlow 1.5.0 on a K80 GPU on Python 2.7

Failing test case:
```python
from __future__ import print_function
import hashlib
import numpy as np
import tensorflow as tf

np.random.randn(2018)
tf.set_random_seed(2018)

X = np.random.randn(1024, 50).astype(np.float32)
b = tf.get_variable('bias', [50])
z = tf.nn.bias_add(X, b)

grad = tf.gradients(z*z, b)[0]

init_op = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init_op)

    run1 = sess.run(grad)
    run2 = sess.run(grad)

    print(np.all(run1 == run2))
    print(np.max(np.abs(run1 - run2)))

    dohash = lambda X: hashlib.md5(X.tostring()).hexdigest()
    print(dohash(run1))
    print(dohash(run2))
```

Outputs
```
False
6.10352e-05
b489a1d659518b2ae9213f5a21e35df2
187a57d563468e59ba5f9d9cf51ca5cb
```

This bug still exists in master, because the code in master uses the unsafe CUDA atomic floating point add in several places. See https://github.com/tensorflow/tensorflow/blob/abc55107eb7a03fe3d83f95fd5e1b8e4def90826/tensorflow/core/kernels/bias_op_gpu.cu.cc 
If TensorFlow will be ever be fully determinstic, atomic floating point add should never be used (it is inherently non-determinstic due to non-associativity of floating point).

Notably, Keras's `Dense` layer uses `bias_add`, so all networks that use this layer are non-reproducible. This is relevant to https://github.com/keras-team/keras/issues/2280 .

 This can currently be avoided by using `tf.add` instead of `tf.nn.bias_add` at a slightly performance hit. The correct fix would be refactor the `BiasAddGrad` op to use a (deterministic) reduction tree.

edit with issue template fields:
Have I written custom code: Python test case, see above
OS Platform and Distribution: RHEL 7.5 (Linux)
TensorFlow installed from: source
TensorFlow version: 1.5.0
Bazel version: unknown
CUDA/cuDNN version: CUDA 8.0.61, cuDNN v6
GPU model and memory: Nvidia K80, 12GB memory
Exact command to reproduce: run the above script
Mobile device: NA

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by wt-huang** [comment]
@eamartin A small amount of non-determinstism in TensorFlow computations is expected on GPU. Ongoing efforts are put forth and FR is in place. [18096](https://github.com/tensorflow/tensorflow/issues/18096) Latest developments from algorithms and hardware architecture perspectives reaped success where challenges remain.

Please resort to workarounds, e.g. tf.add for now until more concrete solutions come forth.

**Comment by wt-huang** [comment]
Closing this issue for now, will use FR to track progress.

**Comment by eamartin** [comment]
@wt-huang This issue is completely separate from https://github.com/tensorflow/tensorflow/issues/18096 as it has nothing to do with cuDNN. The non-determinism here is in the TensorFlow source itself, not a 3rd party library.

**Comment by wt-huang** [comment]
@eamartin This problem is a subset of the non-determinism issue resides in cuDNN. When the feature request is completed and the solution implemented will be transferred to `tf.nn.bias_add`

**Comment by eamartin** [comment]
This problem is a non-determinism issue, but is separate from non-determinism issues with cuDNN.

The solution to cuDNN non-determinism is to pass the correct flags into the library specifying which algorithm to use. See https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility

The solution of "set the right cudDNN flags" cannot be transferred to non-cuDNN issues such as this. Resolving this issue involves modifying CUDA kernels in the TF source. Resolving https://github.com/tensorflow/tensorflow/issues/18096 probably does not.

Additionally, this issue affects more users than https://github.com/tensorflow/tensorflow/issues/18096 as `bias_add` is used for fully connected layers (including through Keras) while cuDNN primarily affects CNNs.

An appropriate "remove non-deterministic behavior" parent issue should not be solely about cuDNN in my opinion. It should include mention of removing all uses of atomic floating point instructions in the TF source.

**Comment by eamartin** [comment]
https://github.com/tensorflow/tensorflow/issues/2732 is a more relevant issue than https://github.com/tensorflow/tensorflow/issues/18096

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by wt-huang** [comment]
@eamartin Yes, atomic floating point addition is one of the primary triggers of non-determinism, which happens to be prevalent in CUDA regime. Often times they go hand it hand. [#2732](https://github.com/tensorflow/tensorflow/issues/2732) is related to this issue so are a few others.

We are very well aware of this issue and progress is underway.

**Comment by eamartin** [comment]
@wt-huang Has this issue been solved? Why has it been closed?

**Comment by wt-huang** [comment]
@eamartin We are aware of this issue, fix is in the progress.

**Comment by eamartin** [comment]
Can this issue remain open until the fix is merged?

**Comment by duncanriach** [comment]
@eamartin, this issue was resolved, in Oct 2019, with PR [31465](https://github.com/tensorflow/tensorflow/pull/31465). In TensorFlow version 2.1 and onwards, you can enable the solution by setting the environment variable `TF_DETERMINISTIC_OPS` to "1" or "true".

For more information, see [github.com/NVIDIA/tensorflow-determinism].

**Comment by eamartin** [comment]
@duncanriach thank you very much for that fix as well as all of your other work on TensorFlow determinism!

**Comment by duncanriach** [comment]
You're welcome, @eamartin. Thank you for the appreciation.
